File: order_service_producer/src/main/java/com/github/marceloedudev/domain/errors/base/NotificationErrors.java
package com.github.marceloedudev.domain.errors.base;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

// https://martinfowler.com/articles/replaceThrowWithNotification.html
public class NotificationErrors {

    private List<AppError> errors = new ArrayList<>();

    public void addError(String message) {
        errors.add(new AppError(message));
    }

    public boolean hasErrors() {
        return !errors.isEmpty();
    }

    public List<String> getErrorsMessage() {
        return errors.stream()
                .map(e -> e.getMessage())
                .collect(Collectors.toList());
    }

    public List<AppError> getErrors() {
        return errors;
    }
}

File: catalog_service/domain/NotificationErrors.go
package domain

import "errors"

type NotificationErrors []error

func (h *NotificationErrors) AddError(e string) {
	*h = append(*h, errors.New(e))
}

func (h *NotificationErrors) HasErrors() bool {
	return len(*h) > 0
}

func (h *NotificationErrors) GetErrorsMessage() []string {
	var errors []string
	for _, e := range *h {
       errors = append(errors, e.Error())
    }
	return errors
}
